src/http.cc

```diff
@@ -1949,6 +2092,48 @@ HttpStateData::httpBuildRequestHeader(HttpRequest * request,
     strConnection.clean();
 }
 
+void
+HttpStateData::makeUpgradeHeaders(HttpHeader &hdr_out)
+{
+    const HttpHeader &hdr_in = request->header;
+
+    if (!hdr_in.has(Http::HdrType::UPGRADE))
+        return;
+
+    if (!Config.http_upgrade_protocols)
+        return;
+
+    const String upgradeIn = hdr_in.getList(Http::HdrType::UPGRADE);
+    String upgradeOut;
+    const char *pos = nullptr;
+    const char *item = nullptr;
+    int ilen = 0;
+    while (strListGetItem(&upgradeIn, ',', &item, &ilen, &pos)) {
```

Comment:
I am not insist that a set of classes to handle upgrade protocols can be useful. Personally I would implement with a different way such classes, but this is does not play any role. 
In any case such classes will increase a lot the code, its complexity and the final performance without gaining anything because these classes will not be used anywhere else.

Also we had a long discussion about the need of a very simple new comparator class  and the complexity it adds instead of using in-lined code.  Is it normal to replace it with something more complex?

Reply:
> In any case such classes will increase a lot the code, its complexity 

It is arguable that the number of lines automatically increases complexity: IMO, a clear design overweighs such kind of complexity. I think (and you agree with it) that good object-oriented design is preferable since it allows to represent existing concepts with classes. I made this (and another) proposal since I noticed several such concepts (such as ProtocolVersion). On the other hand, you may argue that Squid code often violates this approach and the method in question goes in line with may other methods. Your choice.

Reply:
I am not saying that my existing code is optimal, but I do not like also your proposal about the ProtocolVersions, I do not like a super class for all the ProtocolVersions related jobs here. Actually there are two completely different cases:
 - In one case we want to iterate over an strList and extract a full list of protocols without their versions and check against a list of protocols without their version.
- In the second case we need to iterate over an strList and extract a list of protocols with their versions and check against configured acl checks.

The only common they have the above cases is that they refer to "protocols", I am not consider it as a good practice to put them under the same class, with the way you are proposing (a super constructor build the list) and specially if we are not going to use them anywhere else.

Probably a new strListIterate method which accepts functors or lamdas can be used to iterate over list items and do some jobs, and this is will be useful as strList API function. We can implement the two class functors which implements our logic.
Also a simple strListExtractMembers which returns a std::vector of SBuf using a lamda as filter can help us to easy extract lists of protocols to compare, but will add a small performance penalty. Of cource the strListIterate can be a more general solution. 
Note that these two methods I am proposing can be used by other squid subsystems too.

There are other alternatives too.

 I do not know if it worth to try change these two "strListGetItem" loops, but I do not believe that they are so complex.